src/inmanta/execute/scheduler.py

```diff
@@ -430,3 +470,220 @@ def run(self, compiler: "Compiler", statements: Sequence["Statement"], blocks: S
             raise RuntimeException(stmt.expression, "not all statements executed %s" % all_statements)
 
         return True
+
+
+@dataclass(frozen=True)
+class EntityRelationship:
+    """
+    Represent a relationship attribute of an entity.
+    """
+
+    fq_entity_name: str
+    relationship_name: str
+
+    @classmethod
+    def from_type_hint(cls, type_hint: "TypeHint") -> Tuple["EntityRelationship", "EntityRelationship"]:
+        first = cls(type_hint.first_type, type_hint.first_relation_name)
+        then = cls(type_hint.then_type, type_hint.then_relation_name)
+        return first, then
+
+    @classmethod
+    def from_delayed_result_variable(cls, drv: DelayedResultVariable[Any]) -> "EntityRelationship":
+        return cls(
+            fq_entity_name=drv.myself.get_type().get_full_name(),
+            relationship_name=drv.attribute.name,
+        )
+
+
+class PrioritisedDelayedResultVariableQueue:
+    """
+    A queue for DelayedResultVariables that is prioritized based on the
+    type hints passed to the Compiler. This queue will return elements
+    in the following order:
+
+    * First return the DelayedResultVariables, which are not an instance of
+      TempListVariable and that do not have an order constraint.
+    * Then return DelayedResultVariables with order constraint.
+      They are returned in an order that is valid with respect to the
+      constraints.
+    * Finally, all TempListVariables are returned.
+
+    TempListVariables is subclass of DelayedResultVariables that is not
```

Comment:
Why last and not first?

Reply:
https://github.com/inmanta/inmanta-core/pull/3842/files#diff-e7d9b287c2e82ef2aa1f4b76c981aca767736791144db695c7770cb13e8cfac0R407

Reply:
I don't think that answers the question. 

If the hinting is 100% accurate, they should come out first, because if they are safe to freeze, they should be frozen?